adage/analyze/management/commands/import_gene_network.py

```diff
@@ -153,31 +164,42 @@ def check_and_import(file_handle, ml_model):
         # If tokens[1] does not match one and only one gene's
         # systematic_name in the database, skip the line.
         try:
-            gene2 = find_gene(tokens[1])
+            gene2 = find_gene(tokens[1], cached_genes)
         except Exception as e:
             logger.warning("Input file line #%d will be skipped: %s",
                            line_index + 1, e)
             continue
 
         # Check whether the triplet (ml_model, gene1, gene2) is unique.
-        if not unique_together(ml_model, gene1, gene2):
+        if check_unique and not unique_together(ml_model, gene1, gene2):
             raise Exception("Input file line #%d: (%s, %s, %s) is not unique "
                             "in the database" % (line_index + 1,
                                                  tokens[0],
                                                  tokens[1],
                                                  ml_model.title))
 
-        # Import the valid data line into the database.
-        Edge.objects.create(mlmodel=ml_model, gene1=gene1, gene2=gene2,
-                            weight=weight)
+        # Save the valid data into batch.
+        records.append(
+            Edge(mlmodel=ml_model, gene1=gene1, gene2=gene2, weight=weight)
+        )
+        if len(records) == BULK_SIZE:  # dump bulk records into database
+            Edge.objects.bulk_create(records)
+            records = []
+    # Don't forget the last bulk:
+    if len(records) > 0:
+        Edge.objects.bulk_create(records)
+        records = []
```

Comment:
Nice. I am sure that `bulk_create` helps a lot.

The only question I have here is what happens if an error is thrown by `bulk_create`? I guess it's no worse than before, since we don't do any exception handling here. Should we consider wrapping things in a transaction so it can be rolled back if there's a failure? I thought we had done that already somewhere but I can't remember where.

Reply:
@mhuyck: If `bulk_create()` throws an exception, it will be caught by the `except` block on line 57, and the transaction will be rolled back too.

Reply:
Yes, it helped a lot!

Reply:
Aha! I *knew* I had seen a transaction wrapper. I just didn't expand enough of the unchanged code in the diff view to see that. 👍